### PR TITLE
ICD: fix expect response count issue

### PR DESCRIPTION
**Pull request by wqx6**
We should not set `ExpectResponse` to `IsResponseExpected()` in event `kChipMsgSentEvent` as `IsResponseExpected()` indicates whether the exchange context is expecting a response. 
We should set `ExpectResponse` to whether the present message is response-expected.

### CODE REVIEW

**Comment by wqx6** [comment]
The expect response count will become wrong if the exchange context sends a non-response-expected message immediately after it sends a  response-expected message.

**Comment by bzbarsky-apple** [comment]
>  if the exchange context sends a non-response-expected message immediately after it sends a response-expected message.

Note that this should never actually happen.  Have you observed this happening @wqx6 ?

File: src/messaging/ExchangeContext.cpp [hunk 1]
```diff
@@ -243,7 +243,7 @@ CHIP_ERROR ExchangeContext::SendMessage(Protocols::Id protocolId, uint8_t msgTyp
 #if CONFIG_DEVICE_LAYER && CHIP_CONFIG_ENABLE_ICD_SERVER
             DeviceLayer::ChipDeviceEvent event;
             event.Type                       = DeviceLayer::DeviceEventType::kChipMsgSentEvent;
-            event.MessageSent.ExpectResponse = IsResponseExpected();
+            event.MessageSent.ExpectResponse = currentMessageExpectResponse;
```
Comment (pidarped) [inline src/messaging/ExchangeContext.cpp:246]: What does this event capture? Do we actually send back-to-back messages on the same exchange without expecting a response? This seems contrary to expected behavior.

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by pidarped** [comment]
The ExchangeContext::SendMessage logic as it stands, allows the caller to send back to back messages that do not expect responses. It disallows sending only when the second message also expects a response in addition to the first message expecting one. This will only work if the receiver on the other side can dispatch pending Acks for messages in order as it receives back-to-back messages. But I am not sure if we have(or allow for) this use-case. If not then I think we need a gate to disallow sending on the Exchange if a message has been sent already, regardless of whether it expects a response or not.

File: src/messaging/ExchangeContext.cpp (hunk 1 above)
Comment (bzbarsky-apple) [reply src/messaging/ExchangeContext.cpp:246]: > Do we actually send back-to-back messages on the same exchange without expecting a response?

Consider the following message traffic:

1. A sends a message to B.
2. B sends response to A that expects a response from A.
3. A re-sends original message to B (if the response was lost, or just came later than the MRP resend timer).
4. B sends standalone ack for the duplicate.

If step 4 happens before the response to step 2 is received, then B is sending a message that expects a response, then on the same exchange sending a message (the standalone ack) that does not expect a response.

**Comment by bzbarsky-apple** [comment]
> This will only work if the receiver on the other side can dispatch pending Acks for messages in order as it receives back-to-back messages. 

Standalone acks are themselves messages.  So this will work as long as that second message is just a standalone ack.

File: src/messaging/ExchangeContext.cpp (hunk 1 above)
Comment (pidarped) [reply src/messaging/ExchangeContext.cpp:246]: Ok, Ya, I see the flow for sending a StandaloneAck. I presumed that this API was being used by the higher protocol layers only, and MRP control messages(StandaloneAcks) were internally handled and sent down the stack. In that case, this fix makes sense.

**Review by pidarped — APPROVED** [review]

**Pull request merged**